### PR TITLE
Module to gather F5 gtm facts

### PR DESCRIPTION
**Pull request by perzizzle**
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

bigip_gtm_facts
##### Ansible Version:

```
ansible 1.9.2
```
##### Summary:

This module can be used to gather details about F5 GTM Wide Ips, Pools and Virtual Servers
##### Example output:

```
{
    "ansible_facts": {
        "wide_ip": {
            "/Common/test.example.com": {
                "lb_method": "LB_METHOD_ROUND_ROBIN",
                "pool": [{
                    "order": 0,
                    "pool_name": "/Common/MN-test.example.com",
                    "ratio": 1
                }, {
                    "order": 1,
                    "pool_name": "/Common/VA-test.example.com",
                    "ratio": 1
                }]
            }
        }
    },
    "changed": false
}
```

### CODE REVIEW

**Comment by perzizzle** [comment]
@caphrim007 This is the module we talked about awhile back, it works in conjunction with bigip_gtm_wide_ip/pool/virtual_server and is modeled in large part on bigip_facts.

**Comment by caphrim007** [comment]
alrighty, i'll have a look-see. Cursory glances suggest some improvements.

The output listed in the PR needs to be documented in a

RETURN = '''
'''

Block in the head of the file.

Also, there is an f5 file in the module_utils that should be used to get access to the bigip soap api as it includes cert validation. Also it can be used to include the f5_module_spec and then add arguments to that.

There are some examples for reference in this module here

https://github.com/F5Networks/f5-ansible/blob/master/library/bigip_partition.py

Lines 102 and 365.

Thanks!

**Comment by perzizzle** [comment]
@caphrim007 Do you know of any modules that only return ansible_facts for a look at their RETURN string? bigip_facts doesn't seem to support this yet and the developing modules documentation doesn't provide much guidance. I committed an initial guess.

I'll update the module to use the helper methods shortly, those weren't available when I originally wrote this.

**Comment by caphrim007** [comment]
@perzizzle the link I provided at line 365 has an example. It's enough to list the return key (ex. "wide_ip") and an example of the values of the key. For example http://docs.ansible.com/ansible/clc_group_module.html

**Comment by perzizzle** [comment]
@caphrim007 I updated the module to use the f5 module_utils. 

My confusion about the RETURN document string is whether or not I should use ansible_facts (which is the actual key that's returned) or wide_ip, virtual_server, and pools which are the items within ansible_facts. The sample string I provided would be the format returned if you did include=wide_ip.

**Comment by caphrim007** [comment]
@perzizzle the latter. That key causes ansible to merge the facts you provide, as JP mentions here http://jpmens.net/2012/07/15/ansible-it-s-a-fact/

**Comment by caphrim007** [comment]
@perzizzle here's a better example of a dictionary being documented

http://docs.ansible.com/ansible/stat_module.html

**Comment by perzizzle** [comment]
@caphrim007 Alright took another shot at a the RETURN string. Do you know of way to see how it will be displayed in the documentation? It passes yamllint but it would be nice to see what it looks like after its been parsed and displayed.

**Comment by caphrim007** [comment]
@perzizzle Yep, there is the ansible-doc command. You can run it on the existing stat module like so

ansible-doc stat

and see the RETURN section for details. For your module there you can run it and specify the module search path. For example, if your module is located in the current working directory's library/ directory, it might look like this

ansible-doc -M library/ bigip_gtm_facts

Hope that helps

**Comment by gregdek** [comment]
Thanks @perzizzle for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by gregdek** [comment]
Thanks @perzizzle for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by perzizzle** [comment]
I updated version added to pass the build, otherwise I think its ready. @caphrim007 @gregdek

**Comment by perzizzle** [comment]
@gregdek @caphrim007 Updated the version added to pass the build but otherwise I believe this is ready to ship.

File: network/f5/bigip_gtm_facts.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+version_added: "2.2"
+author: 'Michael Perzel'
+notes:
+    - "Requires BIG-IP software version >= 11.4"
+    - "F5 developed module 'bigsuds' required (see http://devcentral.f5.com)"
+    - "Best run as a local_action in your playbook"
+    - "Tested with manager and above account privilege level"
+
+requirements:
+    - bigsuds
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py]: The `requirements` is just text, so you can make this clearer by doing

`- "bigsuds >= 11.4"`

File: network/f5/bigip_gtm_facts.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    - re
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py]: `re` is part of the stdlib so doesn't need to be listed in the `requirements`.

File: network/f5/bigip_gtm_facts.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+import re
+
+
+class F5(object):
+    def __init__(self, host, user, password, validate_certs, session=False):
+        self.api = bigip_api(host, user, password, validate_certs)
+        if session:
+            self.start_session()
+    def get_api(self):
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py:123]: Please add a blank line before `def`

File: network/f5/bigip_gtm_facts.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+
+class VirtualServers(object):
+    def __init__(self, api, regex=None):
+        self.api = api
+        self.virtual_servers = api.GlobalLB.VirtualServerV2.get_list()
+        #Look to support as many fields in dict as available not just name, server
+        if regex:
+            self.virtual_servers = [d for d in self.virtual_servers if regex in d['server'] or regex in d['name']]
+    def get_list(self):
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py:156]: Please add a blank line before `def`

File: network/f5/bigip_gtm_facts.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        return self.virtual_servers
+    def get_enabled_state(self):
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py:159]: Please add a blank line before `def`

File: network/f5/bigip_gtm_facts.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        return self.api.GlobalLB.VirtualServerV2.get_enabled_state(self.virtual_servers)
+    def get_object_status(self):
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py:162]: Please add a blank line before `def`

File: network/f5/bigip_gtm_facts.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    )
+    argument_spec.update(meta_args)
+    
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=False
+    )
+
+    if not bigsuds_found:
+        module.fail_json(msg="the python bigsuds module is required")
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py:266]: `The python bigsuds module is required.`

File: network/f5/bigip_gtm_facts.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        #Look to support as many fields in dict as available not just name, server
+        if regex:
+            self.virtual_servers = [d for d in self.virtual_servers if regex in d['server'] or regex in d['name']]
+    def get_list(self):
+        return self.virtual_servers
+    def get_enabled_state(self):
+        return self.api.GlobalLB.VirtualServerV2.get_enabled_state(self.virtual_servers)
+    def get_object_status(self):
+        return self.api.GlobalLB.VirtualServerV2.get_object_status(self.virtual_servers)
+    #def get_ltm_virtual_server(self):
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py]: Please add details to say why there is commented out code or delete

**Comment by gundalow** [comment]
I'll leave the functional review to the Networking experts.

PEP8 is showing a number of issues that it would be good to get addressed

```
pep8 -r --ignore=E501,E221,W291,W391,E302,E251,E203,W293,E231,E303,E201,E225,E261,E241,E402 bigip_gtm_facts.py 
bigip_gtm_facts.py:123:5: E301 expected 1 blank line, found 0
bigip_gtm_facts.py:152:9: E265 block comment should start with '# '
bigip_gtm_facts.py:155:5: E301 expected 1 blank line, found 0
bigip_gtm_facts.py:157:5: E301 expected 1 blank line, found 0
bigip_gtm_facts.py:159:5: E301 expected 1 blank line, found 0
bigip_gtm_facts.py:161:5: E265 block comment should start with '# '
bigip_gtm_facts.py:163:5: E265 block comment should start with '# '
bigip_gtm_facts.py:165:5: E301 expected 1 blank line, found 0
bigip_gtm_facts.py:219:45: E202 whitespace before '}'
bigip_gtm_facts.py:223:50: E202 whitespace before '}'

```

**Comment by gregdek** [comment]
@perzizzle ping, any update here?

**Comment by perzizzle** [comment]
@gregdek I can make the PEP8 updates next week.

**Comment by gregdek** [comment]
ping @perzizzle :)

**Comment by caphrim007** [comment]
@gregdek I'd be happy to take ownership of this PR and submit all fixes since I'm cleaning up the existing ones anyway and have all the necessary test harnesses to validate.

**Comment by gregdek** [comment]
@caphrim007 Sounds good. Ping me personally for a shipit when you're ready, though -- since the module doesn't exist yet, the regular bot workflow doesn't apply.

**Comment by perzizzle** [comment]
@caphrim007 @gregdek Sorry have been slammed at work. I have an update that fixes the pep clean up but I didn't implement the standardize stuff @caphrim007 has been working on. Let me know if you want to commit the pep fixes.

**Comment by caphrim007** [comment]
@perzizzle if you commit the fixes that @gundalow mentioned above, then i will be glad to fix up remaining work. I'm working on the rest of modules as it is.

**Comment by perzizzle** [comment]
@caphrim007 Updated to play nice with PEP8. Ironically, I just updated our bigsuds library and need to support the validate_certs option throughout all our libraries. Thanks for fixing most of the other F5 ones!

**Comment by gregdek** [comment]
So looks like this PR now suffers from the dreaded "yum failure" Travis issue, which indicates "too old to get proper Travis testing". @perzizzle if you can rebase this against latest so it passes Travis, I'm thinking this should be good to go.

**Comment by gundalow** [comment]
Hi,

Looks like this branch needs updating to fix some issues, please see this guide on how to do that cleanly.

http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

**Comment by perzizzle** [comment]
Yeah this got messy. I'll clean it up.

**Comment by gregdek** [comment]
@perzizzle A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by perzizzle** [comment]
@gregdek quick before the merge conflicts appear again :)

**Comment by caphrim007** [comment]
I'll pull this and check it against my bigips here

**Comment by perzizzle** [comment]
@caphrim007 There's been no changes I just pulled in the latest devel so that there weren't any merge conflicts.

**Comment by gundalow** [comment]
Hi,

Just reviewing Networking PRs. What's the plan for this (who's doing what)?

**Comment by caphrim007** [comment]
Hey @gundalow I think the ball is currently with me and reviewing it to make sure it is returning expected results. I can respond today

**Comment by gundalow** [comment]
@caphrim007 That's great thanks. If there is anything for the Ansible Core Team to do feel free to add it to the Networking Meeting agenda (we have a meeting today)  https://github.com/ansible/community/blob/master/MEETINGS.md

File: network/f5/bigip_gtm_facts.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    bigsuds_found = False
+else:
+    bigsuds_found = True
+
+import re
+
+
+class F5(object):
+    def __init__(self, host, user, password, validate_certs, session=False):
+        self.api = bigip_api(host, user, password, validate_certs)
```
Comment (caphrim007) [inline network/f5/bigip_gtm_facts.py:119]: port needs to go here

File: network/f5/bigip_gtm_facts.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+requirements:
+    - bigsuds >= 11.4
+options:
+    server:
+        description:
+            - BIG-IP host
+        required: true
+        default: null
+        choices: []
+        aliases: []
```
Comment (caphrim007) [inline network/f5/bigip_gtm_facts.py:44]: @perzizzle this module needs to support a server_port variable that is part of the f5 module_utils code now. Would you please add that?

**Comment by caphrim007** [comment]
@gundalow and @perzizzle needs_review

File: network/f5/bigip_gtm_facts.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+requirements:
+    - bigsuds >= 11.4
+options:
+    server:
+        description:
+            - BIG-IP host
+        required: true
+        default: null
+        choices: []
```
Comment (gundalow) [inline network/f5/bigip_gtm_facts.py:43]: You don't need to list empty parameters, so you can delete these lines

**Comment by gundalow** [comment]
needs_revision

**Comment by gregdek** [comment]
Thanks @perzizzle for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@perzizzle A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@perzizzle Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gundalow** [comment]
Merge conflict in `.travis.yml` needs resolving (should be simple)

**Comment by gregdek** [comment]
@perzizzle A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by perzizzle** [comment]
@gregdek @gundalow @caphrim007  Has offered to usher this PR through the approval process.

**Comment by caphrim007** [comment]
@gundalow @gregdek this can be closed for now. I will open a PR that supplies this functionality so that we don't need to be bugging others about it. Thanks!

**Comment by gundalow** [comment]
Thabks again for working on this. Just for completeness once you raise the new PR please link to this one in the comments.

**Comment by caphrim007** [comment]
@perzizzle here is the facts module that has been merged https://github.com/ansible/ansible-modules-extras/pull/3232

**Pull request closed without merging**